transport/src/main/java/io/netty/channel/AbstractCoalescingBufferQueue.java

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.channel;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.internal.UnstableApi;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.ArrayDeque;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static io.netty.util.internal.ObjectUtil.checkPositiveOrZero;
+
+@UnstableApi
+public abstract class AbstractCoalescingBufferQueue {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(AbstractCoalescingBufferQueue.class);
+    private final ArrayDeque<Object> bufAndListenerPairs;
+    private int readableBytes;
+
+    public AbstractCoalescingBufferQueue(int initSize) {
+        bufAndListenerPairs = new ArrayDeque<Object>(initSize);
+    }
+
+    /**
+     * Add a buffer to the front of the queue.
+     */
+    public final void addFirst(ByteBuf buf) {
+        incrementReadableBytes(buf.readableBytes());
+        // Listener would be added here, but since it is null there is no need. The assumption is there is already a
+        // listener at the front of the queue, or there is a buffer at the front of the queue, which was spliced from
+        // buf via remove().
+        bufAndListenerPairs.addFirst(buf);
+    }
+
+    /**
+     * Add a buffer to the end of the queue.
+     */
+    public final void add(ByteBuf buf) {
+        add(buf, (ChannelFutureListener) null);
+    }
+
+    /**
+     * Add a buffer to the end of the queue and associate a promise with it that should be completed when
+     * all the buffers bytes have been consumed from the queue and written.
+     * @param buf to add to the tail of the queue
+     * @param promise to complete when all the bytes have been consumed and written, can be void.
+     */
+    public final void add(ByteBuf buf, ChannelPromise promise) {
+        // buffers are added before promises so that we naturally 'consume' the entire buffer during removal
+        // before we complete it's promise.
+        add(buf, promise.isVoid() ? null : (ChannelFutureListener) new DelegatingChannelPromiseNotifier(promise));
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unnecessary cast to "ChannelFutureListener". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1905)